pkg/controllers/machinehydration/controller.go

```diff
@@ -0,0 +1,125 @@
+	corecloudprovider "github.com/aws/karpenter-core/pkg/cloudprovider"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+	machineutil "github.com/aws/karpenter-core/pkg/utils/machine"
+)
+
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider *cloudprovider.CloudProvider
+}
+
+func NewController(kubeClient client.Client, cloudProvider *cloudprovider.CloudProvider) corecontroller.Controller {
+	return corecontroller.Typed[*v1.Node](kubeClient, &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	})
+}
+
+func (c *Controller) Name() string {
+	return "machinehydration"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, node *v1.Node) (reconcile.Result, error) {
+	if node.Spec.ProviderID == "" {
+		return reconcile.Result{}, nil
+	}
+	provisionerName, ok := node.Labels[v1alpha5.ProvisionerNameLabelKey]
+	if !ok {
+		return reconcile.Result{}, nil
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("provider-id", node.Spec.ProviderID, "provisioner", provisionerName))
+	machineList := &v1alpha5.MachineList{}
+	if err := c.kubeClient.List(ctx, machineList, client.Limit(1), client.MatchingFields{"status.providerID": node.Spec.ProviderID}); err != nil {
+		return reconcile.Result{}, err
+	}
+	if len(machineList.Items) > 0 {
+		return reconcile.Result{}, nil
+	}
+	provisioner := &v1alpha5.Provisioner{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: provisionerName}, provisioner); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	if err := c.hydrate(ctx, node, provisioner); err != nil {
+		return reconcile.Result{}, fmt.Errorf("hydrating machine from node, %w", err)
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) hydrate(ctx context.Context, node *v1.Node, provisioner *v1alpha5.Provisioner) error {
+	machine := machineutil.New(node, provisioner)
+	machine.Name = GenerateName(fmt.Sprintf("%s-", provisioner.Name))
+	machine.Spec.StartupTaints = nil // Assume that startupTaints are nil so that they won't be re-applied to the node
+
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("machine", machine.Name))
+
+	// Hydrates the machine with the correct values if the instance exists at the cloudprovider
+	if err := c.cloudProvider.Hydrate(ctx, machine); err != nil {
+		if corecloudprovider.IsMachineNotFoundError(err) {
+			return nil
+		}
+		return fmt.Errorf("hydrating machine, %w", err)
+	}
+	if err := c.kubeClient.Create(ctx, machine); err != nil {
+		if errors.IsAlreadyExists(err) {
+			return nil
+		}
+		return err
+	}
+	logging.FromContext(ctx).Debugf("hydrated machine from node")
+	return nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(controllerruntime.
+		NewControllerManagedBy(m).
+		For(&v1.Node{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10}))
```

Comment:
consider importing at ctrl, so the other one can take the more natural name controller